enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -351,79 +351,67 @@ impl<'a> Handler<'a> {
     pub fn readv(&mut self) -> u64 {
         self.trace("readv", 3);
 
-        let fd = self.aex.gpr.rdi.raw();
-        let trusted = unsafe {
-            from_raw_parts_mut(
-                self.aex.gpr.rsi.raw() as *mut libc::iovec,
-                self.aex.gpr.rdx.raw() as usize,
-            )
-        };
+        let tiovec: &mut [libc::iovec] =
+            unsafe { self.aex.gpr.rsi.as_slice_mut(self.aex.gpr.rdx.into()) };
 
-        // Add up total size of buffers and size of iovec array.
-        let bufsize = trusted.iter().fold(0, |a, e| a + e.iov_len);
-        let iovecsize = size_of::<libc::iovec>() * trusted.len();
-        let size = bufsize + iovecsize;
-
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size as u64) {
-            Err(errno) => return errno as u64,
-            Ok(map) => unsafe { from_raw_parts_mut(map, size as usize) },
+        // Allocate the correct number of iovecs in untrusted memory.
+        let cursor = self.block.cursor();
+        let uiovec: &mut [libc::iovec] = match unsafe { cursor.alloc(tiovec.len()) } {
+            Ok(slice) => slice,
+            Err(_) => return -libc::EMSGSIZE as u64,
         };
 
-        // Split allocated memory into that used by the iovec struct array and that used by its buffers.
-        let (uiovec, ubuffer) = map.split_at_mut(iovecsize);
-
-        // Convert the prefix from a byte slice into an iovec slice.
-        let (_, untrusted, _) = unsafe { uiovec.align_to_mut::<libc::iovec>() };
-        if untrusted.len() != trusted.len() {
-            self.attacked();
-        }
-
-        // Set pointers in unencrypted iovec slice to use the rest of the allocated memory.
-        // The offset is into the buffer area allocated immediately after the iovec struct
-        // array, measured in bytes.
-        let mut offset = 0;
-        for (t, mut u) in trusted.iter_mut().zip(untrusted.iter_mut()) {
-            u.iov_base = ubuffer[offset..].as_mut_ptr() as *mut _;
+        // Update pointers from iovecs to buffers in untrusted memory.
+        let mut size = 0;
+        for (t, u) in tiovec.iter().zip(uiovec.iter_mut()) {
+            size += t.iov_len;
             u.iov_len = t.iov_len;
-            offset += t.iov_len;
+            u.iov_base = match unsafe { cursor.alloc(t.iov_len) } {
+                Ok(base) => base.as_mut_ptr(),
+                Err(_) => return -libc::EMSGSIZE as u64,
+            };
```

Comment:
There is a very subtle bug here. Relying on type inference means `cursor.alloc()` returns a `&mut [c_void]`. But `size_of::<c_void>() == 0`. So you're actually allocating zero bytes here, regardless of the value of `t.iov_len`.

Reply:
Thank you for pointing this out. I hope I fixed it with `alloc::<usize>`.